quinn-proto/src/config.rs

```diff
@@ -160,19 +161,38 @@ impl TransportConfig {
     /// (see [`TransportConfig::mtu_discovery_config`]).
     ///
     /// Must be at least 1200, which is the default, and known to be safe for typical internet
-    /// applications. Larger values are more efficient, but increase the risk of unpredictable
-    /// catastrophic packet loss due to exceeding the network path's IP MTU. If the provided value
-    /// is higher than what the network path actually supports, packet loss will eventually trigger
-    /// black hole detection and bring it down to 1200.
-    ///
-    /// Real-world MTUs can vary according to ISP, VPN, and properties of intermediate network links
-    /// outside of either endpoint's control. Caution should be used when raising this value for
-    /// connections outside of private networks where these factors are fully controlled.
+    /// applications. Larger values are more efficient, but increase the risk of packet loss due to
+    /// exceeding the network path's IP MTU. If the provided value is higher than what the network
+    /// path actually supports, packet loss will eventually trigger black hole detection and bring
+    /// it down to [`TransportConfig::max_guaranteed_udp_payload_size`].
     pub fn initial_max_udp_payload_size(&mut self, value: u16) -> &mut Self {
         self.initial_max_udp_payload_size = value.max(INITIAL_MAX_UDP_PAYLOAD_SIZE);
         self
     }
 
+    pub(crate) fn get_initial_max_udp_payload_size(&self) -> u16 {
+        self.initial_max_udp_payload_size
+            .max(self.max_guaranteed_udp_payload_size)
+    }
+
+    /// The maximum UDP payload size guaranteed to be supported by the network.
+    ///
+    /// Must be at least 1200, which is the default, and lower than or equal to
+    /// [`TransportConfig::initial_max_udp_payload_size`].
+    ///
+    /// Real-world MTUs can vary according to ISP, VPN, and properties of intermediate network links
+    /// outside of either endpoint's control. Extreme care should be used when raising this value
+    /// outside of private networks where these factors are fully controlled. If the provided value
+    /// is higher than what the network path actually supports, the result will be unpredictable and
+    /// catastrophic packet loss, without a possibility of repair. Prefer
+    /// [`TransportConfig::initial_max_udp_payload_size`] together with
+    /// [`TransportConfig::mtu_discovery_config`] to set a maximum UDP payload size that robustly
+    /// adapts to the network.
+    pub fn max_guaranteed_udp_payload_size(&mut self, value: u16) -> &mut Self {
+        self.max_guaranteed_udp_payload_size = value.max(INITIAL_MAX_UDP_PAYLOAD_SIZE);
```

Comment:
preexisting nit: should we replace "udp_payload_size" here and above with "mtu" for consistency/clarity?

Reply:
You mean from `max_guaranteed_udp_payload_size` to `max_guaranteed_mtu`? I have been avoiding the term "MTU" because I find it ambiguous (e.g. it could also mean the max IP datagram size). If you are confident this change will clarify things, then I'm all for it, though. In that case, should we also rename `initial_max_udp_payload_size` to `initial_mtu`, and other usages of `udp_payload_size` throughout the codebase?

Reply:
I agree that there's some risk of confusion here and I'm interested in other perspectives. I think clarifying in the doc comments will help, as will consistent use of "MTU" throughout the codebase/docs. For example, I think the proposed renaming is more consistent with the use of "MTU" in `MtuDiscoveryConfig` and its documentation than the status quo. We should definitely keep `initial_max_udp_payload_size` consistent with whatever we do here.

Also happy to leave this to be resolved in a separate PR, if you just want to get the rest in.

Reply:
I just went throught all usages of `max_udp_payload_size` and pushed a commit to update them. The main changes are:

* `initial_max_udp_payload_size` -> `initial_mtu`
* `max_guaranteed_udp_payload_size` -> `min_guaranteed_mtu` (note the change in the prefix from `max` to `min`, as this is the minimum MTU that is guaranteed to be supported).
* Variables called (or containing) `max_udp_payload_size` -> `mtu`

I left everything related to the `max_udp_payload_size` transport parameter untouched, so we keep using the names from the spec.

What do you think? Feel free to amend the last commit if there are any obvious improvements that would allow this being merged (I won't have time to tweak it anymore till next week).

Reply:
I think that's a solid improvement in readability across the board, yeah :+1: